### ISSUE TITLE
kears.Model memory leak and tf.funtion block the thread

### ISSUE THREAD

**Issue by yangtao121** [keras-team/keras#15076]
I tried to run Keras Model by multiprocessing.Process for RL, it has some problems. The code as belows:

```python
def get_action(self, obs):
    mu, sigma = self.Model(obs)

    dist = tfp.distributions.Normal(mu, sigma)
    action = tf.squeeze(dist.sample(), axis=0)
    prob = tf.squeeze(dist.prob(action), axis=0)

    return action, prob
```

This code will cause memory leak,but it can work in multiprocess. Then I add `tf.function` decorator, it can run in single process, and has no more memory leak,. However, I when I try to run in multiprocess, this function will block the thread.

**Comment by rmothukuru** [comment]
@yangtao121,
Can you please share the complete reproducible code so that we can try it at our end? Thanks!

**Comment by yangtao121** [comment]
> @yangtao121,
> Can you please share the complete reproducible code so that we can try it at our end? Thanks!

I put the in the code.zip.
[code.zip](https://github.com/keras-team/keras/files/6938319/code.zip)

**Comment by yangtao121** [comment]
> @yangtao121,
> Can you please share the complete reproducible code so that we can try it at our end? Thanks!

also, you can see the code in my github [](https://github.com/yangtao121/RL/blob/master/algo/PPO_Multi.py)

**Comment by yangtao121** [comment]
> @yangtao121,
> Can you please share the complete reproducible code so that we can try it at our end? Thanks!

https://github.com/yangtao121/RL/blob/master/algo/PPO_Multi.py

**Comment by yangtao121** [comment]
> @rmothukuru 
> Can you please share the complete reproducible code so that we can try it at our end? Thanks!
The code you find in https://github.com/yangtao121/RL/blob/master/algo/PPO_Multi.py. And thanks to you

**Comment by yangtao121** [comment]
> @rmothukuru 
> Can you please share the complete reproducible code so that we can try it at our end? Thanks!
The code you can find in https://github.com/yangtao121/RL/blob/master/algo/PPO_Multi.py. And very thanks to you.

**Comment by yangtao121** [comment]
> > @rmothukuru
> > Can you please share the complete reproducible code so that we can try it at our end? Thanks!
> > The code you can find in https://github.com/yangtao121/RL/blob/master/algo/PPO_Multi.py. And very thanks to you.

The leak memory can be solved by upgrade the tensorflow version(2.6.0rc2)

**Comment by rmothukuru** [comment]
@yangtao121,
Can you please let us know if we can close this issue as it can be resolved by upgrading it to 2.6.0rc2? Thanks!

**Comment by yangtao121** [comment]
> @rmothukuru 
> Can you please let us know if we can close this issue as it can be resolved by upgrading it to 2.6.0rc2? Thanks!
sorry, I found it dose not work well, after 1 hour, the memory still goes up.

**Comment by qlzh727** [comment]
@yangtao121, could your provide a minimal repro for this issue? The https://github.com/yangtao121/RL/blob/master/algo/PPO_Multi.py doesn't seems to be a standalone one, and also we don't know how the `get_action` function is invoked.

**Comment by yangtao121** [comment]
@qlzh727 
My Tensorflow version is 2.5.0.

This code turns on multithreading(multiprocessing):

threads = [mp.Process(target=self.rolling, args=[env_name]) for _ in range(self.multi_worker_num)]
self.rolling()function as show below:

 def rolling(self, env_name):
        env = gym.make(env_name).unwrapped
        action_queue = Queue()
        worker = Worker(
            env=env,
            env_args=self.env_args,
            hyper_parameter=self.hyper_parameter,
            get_data_queue=action_queue
        )
        policy = Gaussian_policy(output_queue=action_queue)
        critic = Critic()

        for _ in range(self.epochs):
            policy.load_model('data/policy.h5')
            critic.load_model('data/critic.h5')
            self.ROLLING_EVENT.wait()
            worker.update(policy, critic)
            batch = worker.runner()
            self.batches.put(batch)
            if self.batches.qsize() < self.multi_worker_num - 1:
                self.UPDATE_EVENT.wait()
            # if self.roll_flag < self.multi_worker_num:

            else:
                self.roll_flag = 0
                self.UPDATE_EVENT.set()
                self.ROLLING_EVENT.clear()
The worker is used for sampling data, the policy tell the worker how to do, policy give the action by:

    @tf.function
    def get_action(self, obs):
        # tf.print(obs)
        # print(obs)
        mu, sigma = self.Model(obs)

        dist = tfp.distributions.Normal(mu, sigma)
        action = tf.squeeze(dist.sample(), axis=0)
        prob = tf.squeeze(dist.prob(action), axis=0)
        # print(action)

        return action, prob
worker.runner():

    def runner(self):
        # print('start')
        batches = []
        for i in range(self.trajs):
            collector = Collector(observation_dims=self.obs_dims, action_dims=self.act_dims,
                                  episode_length=self.steps)
            state = self.env.reset()
            # print(i)

            for t in range(self.steps):
                state = state.reshape(1, -1)
                print(state)
                action, prob = self.policy.get_action(state)

                action_ = action * 2

                state_, reward, done, _ = self.env.step(action_)
                collector.store(state, action, reward, prob)
                state = state_

                if (t + 1) % self.batch_size == 0 or t == self.steps - 1:
                    observations, reward = collector.get_current_data()
                    value_ = self.critic.get_value(state_.reshape(1, -1))
                    values = self.critic.get_value(observations)

                    gae, target = gae_target(self.gamma, self.lambada, reward, values, value_, done)

                    collector.get_gae_target(gae, target)

            batches.append(collector)

        return batches
I found when I use multiprocessing, the thread will block at action, prob = self.policy.get_action(state) in worker.runner and had no error output. However, when I remove the tf.function above the policy.get_action() , it can work. But remove the tf.function would cause seriously memory leak.

And the complete code can be found in my github.

Thanks

**Comment by yangtao121** [comment]
My Tensorflow version is 2.5.0.

This code turns on multithreading(multiprocessing):

threads = [mp.Process(target=self.rolling, args=[env_name]) for _ in range(self.multi_worker_num)]
self.rolling()function as show below:

 def rolling(self, env_name):
        env = gym.make(env_name).unwrapped
        action_queue = Queue()
        worker = Worker(
            env=env,
            env_args=self.env_args,
            hyper_parameter=self.hyper_parameter,
            get_data_queue=action_queue
        )
        policy = Gaussian_policy(output_queue=action_queue)
        critic = Critic()

        for _ in range(self.epochs):
            policy.load_model('data/policy.h5')
            critic.load_model('data/critic.h5')
            self.ROLLING_EVENT.wait()
            worker.update(policy, critic)
            batch = worker.runner()
            self.batches.put(batch)
            if self.batches.qsize() < self.multi_worker_num - 1:
                self.UPDATE_EVENT.wait()
            # if self.roll_flag < self.multi_worker_num:

            else:
                self.roll_flag = 0
                self.UPDATE_EVENT.set()
                self.ROLLING_EVENT.clear()
The worker is used for sampling data, the policy tell the worker how to do, policy give the action by:

    @tf.function
    def get_action(self, obs):
        # tf.print(obs)
        # print(obs)
        mu, sigma = self.Model(obs)

        dist = tfp.distributions.Normal(mu, sigma)
        action = tf.squeeze(dist.sample(), axis=0)
        prob = tf.squeeze(dist.prob(action), axis=0)
        # print(action)

        return action, prob
worker.runner():

    def runner(self):
        # print('start')
        batches = []
        for i in range(self.trajs):
            collector = Collector(observation_dims=self.obs_dims, action_dims=self.act_dims,
                                  episode_length=self.steps)
            state = self.env.reset()
            # print(i)

            for t in range(self.steps):
                state = state.reshape(1, -1)
                print(state)
                action, prob = self.policy.get_action(state)

                action_ = action * 2

                state_, reward, done, _ = self.env.step(action_)
                collector.store(state, action, reward, prob)
                state = state_

                if (t + 1) % self.batch_size == 0 or t == self.steps - 1:
                    observations, reward = collector.get_current_data()
                    value_ = self.critic.get_value(state_.reshape(1, -1))
                    values = self.critic.get_value(observations)

                    gae, target = gae_target(self.gamma, self.lambada, reward, values, value_, done)

                    collector.get_gae_target(gae, target)

            batches.append(collector)

        return batches
I found when I use multiprocessing, the thread will block at action, prob = self.policy.get_action(state) in worker.runner and had no error output. However, when I remove the tf.function above the policy.get_action() , it can work. But remove the tf.function would cause seriously memory leak.

And the complete code can be found in my github.

Thanks

**Comment by yangtao121** [comment]
@qlzh727 
My Tensorflow version is 2.5.0.

This code turns on multithreading(multiprocessing):

```python
threads = [mp.Process(target=self.rolling, args=[env_name]) for _ in range(self.multi_worker_num)]
```

`self.rolling()`function as show below:

```python
 def rolling(self, env_name):
        env = gym.make(env_name).unwrapped
        action_queue = Queue()
        worker = Worker(
            env=env,
            env_args=self.env_args,
            hyper_parameter=self.hyper_parameter,
            get_data_queue=action_queue
        )
        policy = Gaussian_policy(output_queue=action_queue)
        critic = Critic()

        for _ in range(self.epochs):
            policy.load_model('data/policy.h5')
            critic.load_model('data/critic.h5')
            self.ROLLING_EVENT.wait()
            worker.update(policy, critic)
            batch = worker.runner()
            self.batches.put(batch)
            if self.batches.qsize() < self.multi_worker_num - 1:
                self.UPDATE_EVENT.wait()
            # if self.roll_flag < self.multi_worker_num:

            else:
                self.roll_flag = 0
                self.UPDATE_EVENT.set()
                self.ROLLING_EVENT.clear()
```

The `worker` is used for sampling data, the `policy` tell the `worker` how to do, `policy` give the action by:

```
    @tf.function
    def get_action(self, obs):
        # tf.print(obs)
        # print(obs)
        mu, sigma = self.Model(obs)

        dist = tfp.distributions.Normal(mu, sigma)
        action = tf.squeeze(dist.sample(), axis=0)
        prob = tf.squeeze(dist.prob(action), axis=0)
        # print(action)

        return action, prob
```

`worker.runner()`:

```python
    def runner(self):
        # print('start')
        batches = []
        for i in range(self.trajs):
            collector = Collector(observation_dims=self.obs_dims, action_dims=self.act_dims,
                                  episode_length=self.steps)
            state = self.env.reset()
            # print(i)

            for t in range(self.steps):
                state = state.reshape(1, -1)
                print(state)
                action, prob = self.policy.get_action(state)

                action_ = action * 2

                state_, reward, done, _ = self.env.step(action_)
                collector.store(state, action, reward, prob)
                state = state_

                if (t + 1) % self.batch_size == 0 or t == self.steps - 1:
                    observations, reward = collector.get_current_data()
                    value_ = self.critic.get_value(state_.reshape(1, -1))
                    values = self.critic.get_value(observations)

                    gae, target = gae_target(self.gamma, self.lambada, reward, values, value_, done)

                    collector.get_gae_target(gae, target)

            batches.append(collector)

        return batches
```

I found when I use multiprocessing, the thread will block at ` action, prob = self.policy.get_action(state)` in `worker.runner`  and had no error output. However, when I remove the `tf.function` above the `policy.get_action()` , it can work. But remove the `tf.function` would cause seriously memory leak.

And the complete code can be found in my [github](https://github.com/yangtao121/RL).

Thanks

**Comment by qlzh727** [comment]
Triage notes: I am not sure whether tf or keras officially support the multiprocess execution (have official doc says we do or don't).  @yangtao121, could you provide some context about why the multiprocess is needed in this case?

**Comment by yangtao121** [comment]
@qlzh727 
I try to use PPO algorithm to complete my project. But before train the model, it needs sampling a lot of samples. So I want to use multipleprocess to shorten the time.
